storage/imx6ull-flash/imx6ull-flash-test.c

```diff
@@ -305,6 +316,122 @@ void test_bootrom(void)
 }
 
 
+/* Checks ECC corrections reported by the BCH */
+void test_ecc(void)
+{
+	const unsigned int paddr = 8 * BLOCK_PAGES_CNT + 1; /* Page under testing (can't contain FCB - different configuration) */
+	const unsigned int rawmetasz = 16 + 26;             /* 16B META + 26B ECC16 */
+	const unsigned int rawdatasz = 512 + 23;            /* 512B DATA + 22.75B ECC14 (adding 2 bits for byte alignment) */
+	const unsigned int maxmetaecc = 16;                 /* Max number of bits per meta chunk ECC can correct */
+	const unsigned int maxdataecc = 14;                 /* Max number of bits per data chunk ECC can correct */
+	unsigned int nblock = paddr / BLOCK_PAGES_CNT;
+	unsigned int start = (paddr % BLOCK_PAGES_CNT) * FLASHDRV_PAGESZ;
+	unsigned int seed = time(NULL);
+	flashdrv_meta_t *aux, *saux;
+	flashdrv_dma_t *dma;
+	uint8_t *data, *sdata;
+	int i, err;
+
+	if ((dma = flashdrv_dmanew()) == MAP_FAILED) {
+		printf("dmanew() failed, err: %d\n", -ENOMEM);
+		return;
+	}
+
+	if ((sdata = mmap(NULL, (FLASHDRV_PAGESZ + _PAGE_SIZE - 1) & ~(_PAGE_SIZE - 1), PROT_READ | PROT_WRITE, MAP_UNCACHED, OID_CONTIGUOUS, 0)) == MAP_FAILED) {
+		printf("mmap() failed: %d\n", -ENOMEM);
+		flashdrv_dmadestroy(dma);
+		return;
+	}
+	saux = (flashdrv_meta_t *)(sdata + flashinfo.writesz);
+
+	if ((data = mmap(NULL, (FLASHDRV_PAGESZ + _PAGE_SIZE - 1) & ~(_PAGE_SIZE - 1), PROT_READ | PROT_WRITE, MAP_UNCACHED, OID_CONTIGUOUS, 0)) == MAP_FAILED) {
+		printf("mmap() failed: %d\n", -ENOMEM);
+		munmap(sdata, (FLASHDRV_PAGESZ + _PAGE_SIZE - 1) & ~(_PAGE_SIZE - 1));
+		flashdrv_dmadestroy(dma);
+		return;
+	}
+	aux = (flashdrv_meta_t *)(data + flashinfo.writesz);
+
+	/* Read original page */
+	if ((err = flashdrv_read(dma, paddr, sdata, saux)) < 0) {
+		printf("read() failed: %d\n", err);
+		munmap(sdata, (FLASHDRV_PAGESZ + _PAGE_SIZE - 1) & ~(_PAGE_SIZE - 1));
+		munmap(data, (FLASHDRV_PAGESZ + _PAGE_SIZE - 1) & ~(_PAGE_SIZE - 1));
+		flashdrv_dmadestroy(dma);
+		return;
+	}
+
+	do {
+		srand(seed);
+
+		/* Flip metadata bits (ECC should correct up to 16 bit flips) */
+		printf("[%4u]: adding bit flips to metadata (%u bad bits already present)\n", paddr, saux->errors[0]);
+		if ((err = flip_bits(maxmetaecc - saux->errors[0], nblock, start, start + rawmetasz, 0)) < 0)
+			break;
+
+		/* Flip data bits (ECC should correct up to 14 bit flips) */
+		for (i = 1; i < sizeof(saux->errors); i++) {
+			printf("[%4u]: adding bit flips to data chunk %u (%u bad bits already present)\n", paddr, i, saux->errors[i]);
+			if ((err = flip_bits(maxdataecc - saux->errors[i], nblock, start + rawmetasz + (i - 1) * rawdatasz, start + rawmetasz + i * rawdatasz, 0)) < 0)
+				break;
```

Comment:
```suggestion
			start += rawmetasz;
			if ((err = flip_bits(maxdataecc - saux->errors[i], nblock, start, start + rawmetasz, 0)) < 0)
				break;
```
It will be simpler.

Reply:
For flipping bits in the data chunks we want to start at offset `start + rawmetasz` and than advance to the next chunk by adding `rawdatasz` (so we'd need `start += rawdatasz`).
However, I don't want to modify `start` variable because it's used later on for restoring flipped bits.

Reply:
My mistake. I mixed up the variables. I meant to suggest using addition instead multiplication.
My suggestion:
1. The `start, nblock, seed` variables can be defined as `const` ( I guess they are not changed ).
2. You cane define a variable:
```
unsigned int starttmp;
```
and use it in `for` loop:
```
starttmp = start + rawmetasz;
for (i = 1; i < sizeof(saux->errors); i++) {
	printf("[%4u]: adding bit flips to data chunk %u (%u bad bits already present)\n", paddr, i, saux->errors[i]);
	if ((err = flip_bits(maxdataecc - saux->errors[i], nblock, starttmp, starttmp + rawdatasz, 0)) < 0)
		break;
		}
	starttmp += rawdatasz;
}
```

Reply:
I've added the proposed changes. I think the code becomes more readable. Thanks.